packages/odyssey-react-theme/src/ThemeProvider/utils.ts

```diff
@@ -0,0 +1,40 @@
+/*!
+ * Copyright (c) 2021-present, Okta, Inc. and/or its affiliates. All rights reserved.
+ * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
+ *
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *
+ * See the License for the specific language governing permissions and limitations under the License.
+ */
+
+import type { Theme, PartialTheme } from "./context";
+
+export const reduceTheme = (
+  prevTheme: Theme,
+  nextTheme: PartialTheme
+): Theme => {
+  if (prevTheme === nextTheme) return prevTheme;
+
+  return Object.entries(nextTheme).reduce(
+    (memo, [k, v]) => (typeof v === "object" ? memo : { ...memo, [k]: v }),
+    prevTheme
+  );
+};
+
+export const asCustomProps = (
+  theme: PartialTheme
+): Record<string, string | number> => {
+  return Object.fromEntries(
+    Object.entries(flat(theme)).map(([k, v]) => [`--${k}`, v])
```

Comment:
Im missing something. Why flatten the theme object here? Why is this necessary?

Reply:
good question - the API for component themes (horizontal customizations) is powered by a computed property name that itself contains a nested object of tokens. By flattening those here we can gather all the relevant custom properties we need to create.